Openas Menu and completion for inheritance and slice roots

Tyzeppelin opened the pull request:
A part of my work last week.

The openas menu is "working". there is still a problem with the cast of the adapters.
The completion is working and I'll improve it to include renaming in it.

Commit:
Update Metamodel and grammar, list of EClass in slice.

Change the reference "roots" in the slice part of the
metamodel/grammar from Strings to EClass/QualifiedName.

Commit:
FEATURE: Completion for inheritance and slice roots

I add custom scope providers to filters inheritable languages and
sliceable parts of a model.

Commit:
Add the attribute "fileExtension" the the malange language extension
pointpoint

Commit:
FEATURE: Add the "Open As" menu to the popup menu

This contribution allows you to open your melange-generated languages
with any "compatible" editor.

Commit:
Clarify the code of the openas menu.

Adding static final String instead of plain String in code.

Commit:
Update Melange MM turning  some string attributes into references

This might not b fully operational. I tried to fix the errors induced by
the MM update. But I need (a hero) someone more skilled than me to ensure I didn't
messed up everything.

Commit:
FEATURE: add support for completion over mapping rules

After changing the metamodel, content assist over mapping rules is now
available.
To do so, we subclass the default proposal provider from java.
The trickiest part was to find wat to change to provide the fully
qualified names of our eobjects instead of just their names.

Tyzeppelin commented:
I've just committed a new update of the metamodel. I'm not sure I didn't messed up some part of the language generation.

tdegueul commented on examples/SimpleAspectsProject/plugin.xml:
```diff
@@ -20,7 +20,8 @@
             uri="http://fsm/"
-            id="simpleaspects.Fsm">
+            id="simpleaspects.Fsm"
+            fileExtension="fsm">
```
I guess you did it by hand. Should I automatically generate it now?

tdegueul commented on plugins/fr.inria.diverse.melange.ui/plugin.xml:
```diff
@@ -782,5 +782,42 @@
+            name="Open As">
+         <commandParameter
+               id="exactType"
+               name="name"
```
Shouldn't the name match the id?

tdegueul commented on plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/contentassist/MelangeProposalProvider.java:
```diff
@@ -0,0 +1,104 @@
+import com.google.inject.Inject;
+
+public class MelangeProposalProvider extends AbstractMelangeProposalProvider {
+	private final static Logger log = Logger.getLogger(AbstractJavaBasedContentProposalProvider.class);
```
The Logger should be retrieved for the current class (MelangeProposalProvider)

tdegueul commented on plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/menu/OpenAsHandler.xtend:
```diff
@@ -0,0 +1,99 @@
+
+
+	// Not used. But it may be needed in the future.
+	def String getAdapter(String exactType, String subType) {
```
In the end we won't use it :) MelangeResource all the way

tdegueul commented on plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/menu/OpenAsHandler.xtend:
```diff
@@ -0,0 +1,99 @@
+
+	// @TODO: Need to fix the ClassNotFoundException,
+	// The adapter class is out of scope
+	@SuppressWarnings("rawtypes")
```
This won't be necessary

tdegueul commented on plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/menu/OpenAsHandler.xtend:
```diff
@@ -0,0 +1,99 @@
+	// @TODO: Need to fix the ClassNotFoundException,
+	// The adapter class is out of scope
+	@SuppressWarnings("rawtypes")
+	override getAdapter(Class adapter) {
```
This won't be necessary

tdegueul commented on plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/scope/MelangeScopeProvider.xtend:
```diff
@@ -0,0 +1,116 @@
+    @Inject XtextResourceSetProvider rsProvider
+    @Inject IQualifiedNameProvider prov
+
+    def dispatch IScope getScope(EObject context, EReference reference) {
```
Not used

tdegueul commented on plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/scope/MelangeScopeProvider.xtend:
```diff
@@ -0,0 +1,116 @@
+    def dispatch IScope getScope(Language language, EReference ref){
+        if (ref.featureID == MelangePackage.LANGUAGE__OPERATORS) {
+        	val owning = language.name
+            val basic = super.getScope((language as EObject), ref).allElements
```
Useless cast I guess

Tyzeppelin replied:
~~Needed to force the dispatcher to call the getScope method for EObject. We don't want it to re-call the getScope for melange over and over again.~~
Edit: My bad then. Fixed.

tdegueul commented on plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/scope/MelangeScopeProvider.xtend:
```diff
@@ -0,0 +1,116 @@
+        	val owning = language.name
+            val basic = super.getScope((language as EObject), ref).allElements
+            val filtered = basic.filter[!it.name.toString.contains(owning)]
+            val scope = new SimpleScope(IScope.NULLSCOPE, filtered)
```
Maybe use the factory instead? Scopes::scopeFor(Iterable<? extends EObject> elements)

tdegueul commented on plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/scope/MelangeScopeProvider.xtend:
```diff
@@ -0,0 +1,116 @@
+
+            return scope
+        }
+        return super.getScope(language as EObject, ref)
```
Useless cast I guess

tdegueul commented on plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/scope/MelangeScopeProvider.xtend:
```diff
@@ -0,0 +1,116 @@
+    def dispatch IScope getScope(Merge merge, EReference ref){
+    	if(ref.featureID == MelangePackage.MERGE__MERGED_LANGUAGE) {
+    		val owning = merge.owningLanguage
+    		val basic = super.getScope((owning) as EObject, ref).allElements
```
Useless cast I guess

tdegueul commented on plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/scope/MelangeScopeProvider.xtend:
```diff
@@ -0,0 +1,116 @@
+    		val owning = merge.owningLanguage
+    		val basic = super.getScope((owning) as EObject, ref).allElements
+    		val filtered = basic.filter[!it.name.toString.contains(owning.name)]
+    		val scope = new SimpleScope(IScope.NULLSCOPE, filtered)
```
Maybe use the factory instead? Scopes::scopeFor(Iterable<? extends EObject> elements)

tdegueul commented on plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/scope/MelangeScopeProvider.xtend:
```diff
@@ -0,0 +1,116 @@
+	def dispatch IScope getScope(Import imp, EReference ref){
+		if (ref.featureID == MelangePackage.IMPORT__MAPPING_RULES) {
+            val genmodelURI = imp.ecoreUri
+			println("On dit ouiche lorraine ")
```
This one is tricky. Either you load the target Ecore file, or you leave it empty and we'll complete it later using some cache mechanism.

tdegueul commented on plugins/fr.inria.diverse.melange/src/main/java/fr/inria/diverse/melange/Melange.xtext:
```diff
@@ -62,19 +62,19 @@ Mapping returns Mapping:
 
 ClassMapping returns ClassBinding:
-	from=STRING "to" to=STRING "{"
+	from=[ecore::EClass | QualifiedName] "to" to=STRING "{"
```
Actually, we may want to rename EClassifiers (including enums & datatypes). Not a big deal though.

tdegueul commented on plugins/fr.inria.diverse.melange/src/main/java/fr/inria/diverse/melange/processors/ExtensionPointProcessor.xtend:
```diff
@@ -116,6 +118,7 @@ class ExtensionPointProcessor extends DispatchMelangeProcessor
 								setAttribute("id", fqn)
 								setAttribute("exactType", l.exactType.fullyQualifiedName.toString)
 								setAttribute("uri", l.syntax.pkgs.head.nsURI)
+								setAttribute("fileExtension", l.syntax.genmodels.head.genPackages.head.fileExtension)
```
Forget my other comment, you did it, great!

tdegueul commented on plugins/fr.inria.diverse.melange/src/main/java/fr/inria/diverse/melange/processors/LanguageBuilder.xtend:
```diff
@@ -325,12 +325,12 @@ class LanguageBuilder extends DispatchMelangeProcessor{
 		
 		rootPackage.EClassifiers.filter(EClass).forEach[cl|
-			if(classes.contains(cl.name)) res.add(cl)
+			if(classes.contains(cl)) res.add(cl)
```
contains() is based on equals() (reference equality), so this probably won't work. A simple name matching should do the job.

Commit:
clean proposal provider.

Commit:
FIX: languageBuilder

"contains() is based on equals() (reference equality), so this probably
won't work. A simple name matching should do the job."

Commit:
Update Metamodel: ClassBinding over EClassifier

Make the classbinding more generic by extending it to all the
Eclassifiers.

Commit:
Name and id of a parameter can be different but

It's easier to give them the same id and name.

Tyzeppelin commented:
__TODO__: Fix a little bug in the menu. The `openEditor()` method don't call the `getURI()` method of our `MelangeEditorInput` class.

Commit:
Open as is now opening file from a "melange:/" URI

It still can't display the file properly due to problems with adapters.

Commit:
Merge conflict

Commit:
BUGFIX: from.name in languagebuilder

Commit:
Typo: bad filename
